File: sol/1492.cpp
struct Solution {
    int kthFactor(int n, int k) {
        vector<int> small, large;
        for (int i = 1; i * i <= n; ++i){
            if (n % i == 0){
                small.push_back(i);
                int q = n / i;
                if (q != i){
                    large.push_back(q);
                }
                if (small.size() == k){
                    return i;
                }
            }
        }
        if (small.size() + large.size() >= k){
            return large[large.size() + small.size() - k];
        } else {
            return -1;
        }
    }
};

File: 1492.py
class Solution:
    def kthFactor(self, n: int, k: int) -> int:
        small = []
        large = []
        i = 1
        while i ** 2 <= n:
            q, r = divmod(n, i)
            if r == 0:
                small.append(i)
                if i != q:
                    large.append(q)
                if len(small) == k:
                    return i
            i += 1
        if (idx := len(small) + len(large) - k) >= 0:
            return large[idx]
        else:
            return -1
